dmoj/cli.py

```diff
@@ -188,33 +188,36 @@ def execute(self, line):
         args = self.arg_parser.parse_args(line)
 
         problem_id = args.problem_id
+        language_id = args.language_id
         time_limit = args.time_limit
         memory_limit = args.memory_limit
-        language_id = args.language_id
+        source_file = args.source_file
+
+        if language_id not in executors:
+            source_file = language_id
+            language_id = None    # source file / language id optional
 
         err = None
 
         if problem_id not in map(itemgetter(0), judgeenv.get_supported_problems()):
             err = "unknown problem '%s'" % problem_id
-        elif language_id == 'unknown':
-            if args.source_file:
-                ext = None
-                try:
-                    filename, ext = args.source_file.split('.')
-                except Exception:
+        elif not language_id:
+            if source_file:
+                filename, dot, ext = source_file.partition('.')
+                if not ext:
                     err = 'invalid file name'
-                ext = ext.upper()
-                print ext
-                if ext == 'PY':
-                    language_id = 'PY2'
-                elif ext == 'CPP':
-                    language_id = 'CPP14'
-                elif ext == 'JAVA':
-                    language_id = 'JAVA9'
-                elif ext:
-                    language_id = ext
                 else:
-                    err = "unknown language '%s'" % language_id
+                    ext = ext.upper()
+                    if ext == 'PY':
+                        language_id = 'PY2'
+                    elif ext == 'CPP':
+                        language_id = 'CPP14'
```

Comment:
CPP11 here.

In theory we ought to have this use the latest version of CPP available, and the latest version of Java available etc., but in practice I doubt that'd be worth the extra code complexity (this tool is intended mostly for developers, who likely know what they're doing).